solr/core/src/java/org/apache/solr/cluster/events/ScheduledEvent.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.cluster.events;
+
+/**
+ *
+ */
+public interface ScheduledEvent extends ClusterEvent {
```

Comment:
Sorry, this was asked here already: https://github.com/apache/lucene-solr/pull/1758#discussion_r471838373

Reply:
Leader rebalancing job, cluster rebalancing, removing replicas from a node to be removed, unnecessary replica removal, core corruption checking etc etc

Reply:
Right, generally speaking any maintenance or optimization task that needs to be periodically invoked. Again, we could define it as a separate API but I think it fits the event model nicely.

Reply:
> Leader rebalancing job, cluster rebalancing, removing replicas from a node to be removed, unnecessary replica removal, core corruption checking etc etc

AFAICT, these should be actions/operations that should be plugged into the autoscaling framework package/module. And, the scheduling for these should also be completely self-contained within that framework (with no traces in the solr-core). IOW, ScheduledEvent should not be a first class concept/interface in "solr-core".

Reply:
There is nothing wrong in making all these additions in a module.

Please not that this module is a part of Solr. it will be shipped with Solr and the user may just have to run a command to enable that

Reply:
> plugged into the autoscaling framework
This PR is NOT just about the autoscaling. The goal is to define a generic API for cluster-level events. Autoscaling plugins will be just one consumer of this API.

> IOW, ScheduledEvent should not be a first class concept/interface in "solr-core"
I disagree. Scheduling various maintenance tasks is a common need (not just in autoscaling), and it fits with the event API. Sure, we can define a separate API just for scheduling, but it would be still something that belongs to solr-core.

Reply:
I disagree that scheduling should be a first class concept in Solr. Users/devops can (and currently do) use their own schedulers to carry out maintenance tasks for Solr. But, if Solr can provide this functionality as a "scheduler-module" or "scheduler-plugin", I am fine. But, I see no reason why scheduling should be a solr-core concept.

Reply:
> Leader rebalancing job, cluster rebalancing, removing replicas from a node to be removed, unnecessary replica removal, core corruption checking etc etc

For some context, almost all devops engineers I've interacted with do not trust Solr to carry out these things in an unsupervised manner because of potential disruption to stability because of these "maintenance operations".

Reply:
Is the package system mature enough to handle a dependency graph? I.e 3rd party pks A ---> 1st party pkg B ---> 1st party package C? Would attempt to install A give an error that B must be installed first? Or would install of A trigger auto install of (1st party) pkg B?

I'd prefer to not go there yet, and let packages depend on core APIs only as of now.

We could perhaps make schedulers pluggable, and only include a very slim `TimeIntervalScheduler` in core, and let plugins add their more complex schedulers such as `CronScheduler` as part of the package if needed?

Reply:
Even if package manager doesn't support dependencies today, this can still be loaded from contrib modules (that produces separate jar) that depend on reach other. Modularization is the first step towards package manager based loading. If we start with everything in solr core, then it will be considerably harder to make them into packages.

Reply:
First step is separating interfaces and implementation. If `solr-core` is to interact with non core modules, it needs a way to talk to them. After interfaces comes the question of instantiation or dependency injection into `solr-core`.
Even if package selection/install is not automated, having the right code for a clean separation will make evolution much easier.
To some extent, if `solr-core` doesn't need to talk (directly or indirectly) to another module, then maybe the module doesn't belong at all in Solr?

Reply:
> For some context, almost all devops engineers I've interacted with do not trust Solr to carry out these things in an unsupervised manner

IIUIC this objection was to the scheduler part. This is now removed from this PR - changes here are related purely to the facility to generate cluster-state related events and the API for registering listeners to these events.